drivers/can/can_renesas_ra.c

```diff
@@ -0,0 +1,1171 @@
+void can_renesas_ra_fsp_cb(can_callback_args_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+
+	switch (p_args->event) {
+	case CAN_EVENT_RX_COMPLETE: {
+		struct can_frame z_frame = {0};
+
+		FSP_TO_ZEPHYR_FRAME(&p_args->frame, &z_frame);
+		can_renesas_ra_call_rx_cb(dev, &z_frame);
+		break;
+	}
+
+	case CAN_EVENT_TX_COMPLETE: {
+		can_renesas_ra_call_tx_cb(dev);
+		break;
+	}
+
+	case CAN_EVENT_ERR_CHANNEL: {
+		if (p_args->error & R_CANFD_CFDC_ERFL_BEF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_ACTIVE);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_EWF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_WARNING);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_EPF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_PASSIVE);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_BOEF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_BUS_OFF);
+		}
+		break;
+	}
+
+	default:
+		break;
+	}
+}
+
+static inline void can_renesas_ra_call_tx_cb(const struct device *dev)
+{
+	struct can_renesas_ra_data *data = dev->data;
+
+	data->custom.tx_cb(dev, 0, data->custom.tx_usr_data);
+	k_work_cancel_delayable(&data->custom.tx_timeout_work);
+}
+
+static inline void can_renesas_ra_call_rx_cb(const struct device *dev, struct can_frame *rx_frame)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	const struct can_renesas_ra_cfg *cfg = dev->config;
+	struct can_renesas_ra_filter *rx_filter = data->custom.rx_filter;
+
+	for (uint32_t i = 0; i < cfg->custom.rx_filter_num; i++) {
+		if (!rx_filter->rx_cb) {
+			rx_filter++;
+			continue; /* this filter has not set yet */
+		}
+
+		if (!can_frame_matches_filter(rx_frame, &rx_filter->filter)) {
+			rx_filter++;
+			continue; /* filter did not match */
+		}
+
+		rx_filter->rx_cb(dev, rx_frame, rx_filter->rx_usr_data);
+		break;
+	}
+}
+
+static inline void can_renesas_ra_call_state_change_cb(const struct device *dev,
+						       enum can_state state)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	can_info_t can_info = {};
+	fsp_err_t fsp_err;
+
+	fsp_err = can_api->infoGet(data->custom.fsp_can.p_ctrl, &can_info);
+	if (fsp_err != FSP_SUCCESS) {
+		LOG_DBG("drivers: can: can get err info failed");
+		return;
+	}
+	struct can_bus_err_cnt err_cnt = {
+		.rx_err_cnt = can_info.error_count_receive,
+		.tx_err_cnt = can_info.error_count_transmit,
+	};
+
+	if (data->common.state_change_cb == NULL) {
+		LOG_DBG("drivers: can: state change callback is not set");
+		return;
+	}
+	data->common.state_change_cb(dev, state, err_cnt, data->common.state_change_cb_user_data);
+}
+
+static void can_renesas_ra_tx_timeout_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct can_renesas_ra_custom_data *cus_data =
+		CONTAINER_OF(dwork, struct can_renesas_ra_custom_data, tx_timeout_work);
+
+	cus_data->tx_cb(cus_data->dev, -EAGAIN, cus_data->tx_usr_data);
+}
```

Comment:
I am having a hard time understanding what is going on with this? Is it an attempt to to TX retries in software?

Reply:
I removed this code